R/match_pc_gpprac.R

```diff
@@ -0,0 +1,379 @@
+      dplyr::anti_join(data, lookup, by = "gpprac") %>%
+        dplyr::mutate(gpprac_match = 0L)
+    )
+  }
+
+  return(return_data)
+}
+
+#' Use the mean to find situations where postcodes do not match over
+#' chi groups
+#'
+#' @param data A data frame
+#' @param match_variable Can be postcode_match or gpprac_match
+#'
+#' @return A data frame with variable all_match, which is 0 or 1 if all
+#' postcodes match and a real number between 0 and 1 when they don't
+find_matching_data_by_chi <- function(data, match_variable) {
+  # Create all_match, the mean of the match variable, for those chis that have
+  # some matched and some unmatched
+  return_data <- data %>%
+    dplyr::group_by(chi) %>%
+    dplyr::summarise(all_match = mean({{ match_variable }})) %>%
+    dplyr::ungroup()
+
+  return(return_data)
+}
+
+#' Use situations where some records for a chi have a postcode to fill
+#' those that are missing
+#'
+#' @param data A data frame
+#' @param matching_data A data frame processed through [find_matching_data_by_chi()]
+#' @param type Either pc for pastcodes or gp for GP practices
+#'
+#' @return A data frame with some missing data filled based on type
+fill_data_from_chi <- function(data, matching_data, type = c("pc", "gp")) {
+  if (type == "pc") {
+    return_data <-
+      dplyr::left_join(data,
+        matching_data,
+        by = "chi"
+      ) %>%
+      # If the chi isn't missing and all_match isn't exactly 0 or 1,
+      # we can potentially fill the postcodes for that chi
+      dplyr::filter(!is_missing(chi) & !is.integer(all_match)) %>%
+      dplyr::group_by(chi) %>%
+      # Arrange by one of the keydates so the most recent postcode is at the top even if it's NA
+      # Then fill the values of postcode upwards so the NA is filled in
+      dplyr::arrange(
+        dplyr::desc(is.na(postcode)),
+        dplyr::desc(keydate2_dateformat),
+        .by_group = TRUE
+      ) %>%
+      # "NK010AA" is a known dummy postcode, so we replace this with NA
+      dplyr::mutate(postcode = dplyr::if_else(postcode == "NK010AA",
+        NA_character_,
+        postcode
+      )) %>%
+      # Now we can fill the postcodes for these chis
+      tidyr::fill(postcode, .direction = "up") %>%
+      dplyr::ungroup()
+  } else if (type == "gp") {
+    return_data <- dplyr::left_join(data_gpprac_match, data_for_matching, by = "chi") %>%
+      dplyr::filter(!is_missing(chi) & !is.integer(all_match)) %>%
+      dplyr::group_by(chi) %>%
+      # Arrange by one of the keydates so the most recent gpprac is at the top even if it's NA
+      # Then fill the values of gpprac upwards so the NA is filled in
+      dplyr::arrange(
+        dplyr::desc(is.na(gpprac)),
+        dplyr::desc(keydate2_dateformat),
+        .by_group = TRUE
+      ) %>%
+      dplyr::mutate(gpprac = dplyr::if_else(gpprac == 99999L, NA_real_, gpprac)) %>%
+      tidyr::fill(postcode, .direction = "up") %>%
+      dplyr::ungroup()
+  }
+
+  return(return_data)
+}
+
```

Comment:
I had an idea of how to trim down this function as essentially it is repetitive for the pc and gpprac variables apart from where we are trying to replace the dummy codes with NA. I still need to test this as my R session has been very slow today but i have wrote down my idea in a script:

'''
fill_data_from_chi <- function(data, matching_data, match_var) {

    return_data <-
      dplyr::left_join(data,
                       matching_data,
                       by = "chi"
      ) %>%
      # If the chi isn't missing and all_match isn't exactly 0 or 1,
      # we can potentially fill the postcodes for that chi
      dplyr::filter(!is_missing(chi) & !is.integer(all_match)) %>%
      dplyr::group_by(chi) %>%
      # Arrange by one of the keydates so the most recent postcode is at the top even if it's NA
      # Then fill the values of postcode upwards so the NA is filled in
      dplyr::arrange(
        dplyr::desc(is.na({{match_var}})),
        dplyr::desc(keydate2_dateformat),
        .by_group = TRUE
      ) %>%
      dplyr::mutate({{match_var}} = replace_dummy_variables( {{match_var}} ) %>% 
      # Now we can fill the variables for these chis
      tidyr::fill({{match_var}}, .direction = "up") %>%
      dplyr::ungroup()
      
    return(return_data)
}
'''

Essentially, here i am replacing type with a match_var which would be either "postcode" or "gpprac" and there is a new function for dealing with the dummy codes: 

'''
replace_dummy_variables <- function(match_var = c("postcode", "gpprac")) {
  if (match_var == "postcode") {
    return_data <- dplyr::if_else(postcode == "NK010AA", NA_character_, postcode)
  } else if (match_var == "gpprac") {
    return_data <- dplyr::if_else(gpprac == 99999L, NA_real_, gpprac)
  }

  return(return_data)
}
'''

I still need to test this is working but this may help with trimming down the code to produce the same result